lib/howl/command.moon

```diff
@@ -182,48 +83,147 @@ command_completer = ->
   matcher = Matcher items
   (text) -> matcher(text), completion_options
 
-direct_dispatch = (cmd_string) ->
-  return false if not cmd_string or cmd_string.is_blank
-  cmd, cmd_name, text = parse_cmd cmd_string
-  cmd or= resolve_command cmd_string
-  return false unless cmd
-  input = load_input cmd.input, text
-  return false if input
-  cmd!
-  true
-
-run = (cmd_string = nil) ->
-  return if direct_dispatch cmd_string
-
-  cmd_completer = nil
-  state = State!
-
-  cmd_input =
-    title: 'Command'
-    should_complete: (_, readline) -> state\should_complete readline
-    close_on_cancel: (_, readline) -> state\close_on_cancel readline
-    update: (_, text, readline) -> state\update text, readline
-    on_selection_changed: (_, item, readline) -> state\on_selection_changed item, readline
-    on_completed: (_, item, readline) -> state\on_completed item, readline
-    on_cancelled: (_, readline) -> state\on_cancelled readline
-    on_submit: (_, text, readline) -> state\on_submit text, readline
-    go_back: (_, readline) -> state\go_back readline
-    keymap: state.keymap
-
-    complete: (_, text, readline) ->
-      if state.cmd
-        return state\complete(text, readline)
+get_command_items = ->
+  cmd_names = names!
+  bindings = command_bindings!
+
+  table.sort cmd_names
+  items = {}
+  for name in *cmd_names
+    def = commands[name]
+    desc = def.description
+    if def.alias_for
+      desc = "(Alias for #{def.alias_for})"
+      desc = "[deprecated] #{desc}" if def.deprecated
+    binding = bindings[name] or ''
+    append items, { name, binding, desc }
+
+  return items
+
+class CommandRunner
+  run: (@finish, opts={}) =>
+    @command_line = howl.app.window.command_line
+    cmd = nil
+    cmdname = ''
+    @command_line.prompt = ':'
+    @command_line.title = 'Command'
+    if opts.directory
+      @command_line.directory = opts.directory
+    if opts.cmd_string
+      @run_command opts.cmd_string
+
+  run_command: (cmd_string) =>
+    cmd, cmdname, text = parse_cmd cmd_string
+
+    if cmd
+      @command_line\write_spillover text
+    else
+      cmd = resolve_command cmd_string
+      cmdname = cmd and cmd.name
+
+    if not cmd
+      @command_line.text = cmd_string
+      log.error 'No such command'
+      return
+
+    @command_line\clear!
+    @command_line.prompt = markup.howl "<prompt>:</><commandname>#{cmdname}</> "
+
+    local results
+
+    ok, error = pcall ->
+      results = table.pack howl.app.window.command_line\run cmd
+
+    if not ok
+      log.error error
+
+    self.finish results and unpack results
+
+  on_update: (text) =>
+    if text\find ' '
```

Comment:
This line itself is invoked only when you are in the CommandRunner interaction - IOW, when the completion list is not open. E.g. if you just type 'open file' without first hitting `tab`, this line will cause the open command to run as soon as you type the space. The arguments after that are handled by the invoked command and may have spaces (depending on what the command allows).

The undesirable behavior you describe is from using submit_on_space when calling `interact.select` below. The selection list interaction has no knowledge of commands, it just matches items in the list and submits when you hit `space`. That behavior can be tweaked, it would require rolling the completion list into the CommandRunner rather than running it as a nested interaction, since we want more control that what `interact.select` provides. 

However consider what behavior you want if you type space with the list showing - should it search the list for two words or should it run a command? E.g. in the current version of howl, with the command completion list showing, if I type 'new buffer', I see that it matches the description of a command. However if I type 'close buffer', it immediately runs the close command rather than show me matching commands.